swarm/network/simulations/overlay.go

```diff
@@ -72,162 +70,27 @@ func (s *Simulation) NewService(ctx *adapters.ServiceContext) (node.Service, err
-			panic(err.Error())
-		}
-	}
-	for i, id := range ids {
-		log.Trace(fmt.Sprintf("setup mocker: register a peer on node %x", id[:4]))
-		var peerID discover.NodeID
-		if i == 0 {
-			peerID = ids[len(ids)-1]
-		} else {
-			peerID = ids[i-1]
-		}
-		ch := make(chan network.OverlayAddr)
-		go func() {
-			defer close(ch)
-			ch <- network.NewAddrFromNodeID(peerID)
-		}()
-		log.Trace(fmt.Sprintf("%x registers peer %x", id[:4], peerID[:4]))
-		if err := net.GetNode(id).Node.(*adapters.SimNode).Services()[0].(*network.Bzz).Hive.Register(ch); err != nil {
-			panic(err.Error())
-		}
-	}
-
-	return ids
-}
-
-func bootMocker(net *simulations.Network) {
-	setupMocker(net)
-}
-
-func randomMocker(net *simulations.Network) {
-	ids := setupMocker(net)
-
-	for {
-		var lowid, highid int
-		var wg sync.WaitGroup
-		randWait := rand.Intn(5000) + 1000
-		rand1 := rand.Intn(9)
-		rand2 := rand.Intn(9)
-		if rand1 < rand2 {
-			lowid = rand1
-			highid = rand2
-		} else if rand1 > rand2 {
-			highid = rand1
-			lowid = rand2
-		} else {
-			if rand1 == 0 {
-				rand2 = 9
-			} else if rand1 == 9 {
-				rand1 = 0
-			}
-			lowid = rand1
-			highid = rand2
-		}
-		var steps = highid - lowid
-		wg.Add(steps)
-		for i := lowid; i < highid; i++ {
-			log.Info(fmt.Sprintf("node %v shutting down", ids[i]))
-			net.Stop(ids[i])
-			go func(id discover.NodeID) {
-				time.Sleep(time.Duration(randWait) * time.Millisecond)
-				net.Start(id)
-				wg.Done()
-			}(ids[i])
-			time.Sleep(time.Duration(randWait) * time.Millisecond)
-		}
-		wg.Wait()
-	}
-}
-
-func startStopMocker(net *simulations.Network) {
-	ids := setupMocker(net)
-
-	for range time.Tick(10 * time.Second) {
-		id := ids[rand.Intn(len(ids))]
-		go func() {
-			log.Error("stopping node", "id", id)
-			if err := net.Stop(id); err != nil {
-				log.Error("error stopping node", "id", id, "err", err)
-				return
-			}
-
-			time.Sleep(3 * time.Second)
-
-			log.Error("starting node", "id", id)
-			if err := net.Start(id); err != nil {
-				log.Error("error starting node", "id", id, "err", err)
-				return
-			}
-		}()
-	}
-}
-
 // var server
 func main() {
 	flag.Parse()
 
 	runtime.GOMAXPROCS(runtime.NumCPU())
 
-	log.Root().SetHandler(log.LvlFilterHandler(log.LvlTrace, log.StreamHandler(os.Stderr, log.TerminalFormat(false))))
+	log.Root().SetHandler(log.LvlFilterHandler(log.LvlInfo, log.StreamHandler(os.Stderr, log.TerminalFormat(true))))
 
 	s := NewSimulation()
 	services := adapters.Services{
 		"overlay": s.NewService,
 	}
 	adapter := adapters.NewSimAdapter(services)
-
 	network := simulations.NewNetwork(adapter, &simulations.NetworkConfig{
 		DefaultService: "overlay",
 	})
 
-	mockers := createMockers()
-
-	config := simulations.ServerConfig{
-		DefaultMockerID: "randomNodes",
-		// DefaultMockerID: "bootNet",
-		Mockers: mockers,
-	}
+	nodeCount := 30
+	log.Info("starting simulation node runner on 0.0.0.0:8889...")
+	go simulations.NewNodeRunner(network, "probabilistic", nodeCount)
```

Comment:
Does this need to be a separate HTTP server rather than just being part of the existing API?

Reply:
No, it doesn't need to be - my impression was that this was all going to be kept out of `network.go`, that's why the separation, but it doesn't make much sense indeed...so you vote to move it to existing API?

Reply:
So the original suggestion was to move this logic out of the backend and into the client, but given you want to keep it in the backend then we may as well just keep it in the existing HTTP API.

I think the API should look like this:

* `POST /mocker/start` - starts the mocker, with algorithm type (e.g. `start-stop` `probabilistic` etc.) in the body / query param

* `POST /mocker/stop` - stops the currently running mocker

The server should have the following fields:

```go
type Server struct {
	mockerStop chan struct{} // when set, stops the current mocker
	mockerMtx  sync.Mutex    // synchronises access to the mockerStop field
}
```

Then starting a mocker looks something like:

```go
srv.mockerMtx.Lock()
defer srv.mockerMtx.Unlock()
if srv.mockerStop != nil {
	return errors.New("mocker already running")
}
mockerType := req.FormValue("mocker-type")
mockerFn := lookupMocker(mockerType)
if mockerFn == nil {
	return fmt.Errorf("unknown mocker type %q", mockerType)
}
srv.mockerStop = make(chan struct{})
go mockerFn(srv.network, srv.mockerStop)
```

and stopping it:

```go
srv.mockerMtx.Lock()
defer srv.mockerMtx.Unlock()
if srv.mockerStop == nil {
	return nil
}
close(srv.mockerStop)
srv.mockerStop = nil
```

A `mockerFn` has a signature `func(*Network, chan struct{})`, for example:

```go
func startStop(net *Network, quit chan struct{}) {
	tick := time.NewTicker(10*time.Second)
	defer tick.Stop()
	for {
		select {
		case <-tick.C:
			// start / stop nodes
		case <-quit:
			return
		}
	}
}
```

Reply:
Ok, implemented something accordingly.